cpp/src/arrow/testing/random.cc

```diff
@@ -772,12 +789,42 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(const Field& field, int64_t
     }
 
       GENERATE_INTEGRAL_CASE_VIEW(Int32Type, Date32Type);
-      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, Date64Type);
       GENERATE_INTEGRAL_CASE_VIEW(Int64Type, TimestampType);
-      GENERATE_INTEGRAL_CASE_VIEW(Int32Type, Time32Type);
-      GENERATE_INTEGRAL_CASE_VIEW(Int64Type, Time64Type);
       GENERATE_INTEGRAL_CASE_VIEW(Int32Type, MonthIntervalType);
 
+    case Type::type::DATE64: {
+      using c_type = typename Date64Type::c_type;
+      constexpr c_type kFullDayMillis = 1000 * 60 * 60 * 24;
+      constexpr c_type min_value = std::numeric_limits<c_type>::min() / kFullDayMillis;
+      constexpr c_type max_value = std::numeric_limits<c_type>::max() / kFullDayMillis;
+
+      return *Numeric<Date64Type>(length, min_value, max_value, null_probability)
+                  ->View(field.type());
+    }
+
+    case Type::type::TIME32: {
+      TimeUnit::type unit = std::dynamic_pointer_cast<Time32Type>(field.type())->unit();
```

Comment:
nit: we have `internal::checked_[pointer_]cast` which will use dynamic or static cast depending on debug/release build